src/aes_xts.c

```diff
@@ -259,6 +260,82 @@ zpc_aes_xts_set_iv(struct zpc_aes_xts *aes_xts, const u8 * iv)
 	return rc;
 }
 
+int
+zpc_aes_xts_get_intermediate_iv(struct zpc_aes_xts *aes_xts, unsigned char iv[16])
+{
+	int rc, off1;
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		goto ret;
+	}
+	if (!hwcaps.aes_xts) {
+		rc = ZPC_ERROR_HWCAPS;
+		goto ret;
+	}
+	if (aes_xts == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		goto ret;
+	}
+
+	if (iv == NULL) {
+		rc = ZPC_ERROR_ARG2NULL;
+		goto ret;
+	}
+
+	if (aes_xts->iv_set != 1) {
+		rc = ZPC_ERROR_IVNOTSET;
+		goto ret;
+	}
+
+	off1 = AES_XTS_KM_XTSPARAM(aes_xts->aes_key1->keysize);
+	memcpy(iv, aes_xts->param_km + off1, 16);
+	rc = 0;
+
+ret:
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int
+zpc_aes_xts_set_intermediate_iv(struct zpc_aes_xts *aes_xts, const unsigned char iv[16])
+{
+	int rc;
+
+	if (pkeyfd < 0) {
+		rc = ZPC_ERROR_DEVPKEY;
+		goto ret;
+	}
+	if (!hwcaps.aes_xts) {
+		rc = ZPC_ERROR_HWCAPS;
+		goto ret;
+	}
+	if (aes_xts == NULL) {
+		rc = ZPC_ERROR_ARG1NULL;
+		goto ret;
+	}
+	if (iv == NULL) {
+		rc = ZPC_ERROR_ARG2NULL;
+		goto ret;
+	}
+
+	if (aes_xts->key_set != 1) {
+		rc = ZPC_ERROR_KEYNOTSET;
+		goto ret;
+	}
```

Comment:
I guess you should also check for aes_xts->iv_set here. Not really required currently, but most likely it will be required in the future. To not beak the API at a later time, lets require it right now already. 
Maybe also mention the requirement for calling this function in the header file comments, i.e. key must have been set and iv (aka tweak) must have been set before setting the intermediate iv.

Reply:
Yes, although not really technically required, but better make that a requirement right now.

We should say that a context must have been fully set up before a intermediate IV can be set, whatever that means for the specific cipher mode.